hydromt/raster.py

```diff
@@ -1697,8 +1707,9 @@ def set_nodata(self, nodata=None, logger=logger):
                 nodata = self._obj.rio.encoded_nodata
         # Only numerical nodata values are supported
         if np.issubdtype(type(nodata), np.number):
-            self._obj.rio.set_nodata(nodata, inplace=True)
-            self._obj.rio.write_nodata(nodata, inplace=True)
+            # cast to float since using int causes inconsistent casting
```

Comment:
what do you mean by this? Some dataarray are of type type or uint and the nodata value should be same type as array values type.

Reply:
Well, rioarray generates this warning:

```
tests/test_model.py::test_gridmodel_setup
  /home/sam/.micromamba/envs/hydromt/lib/python3.11/site-packages/rioxarray/raster_writer.py:130: UserWarning: The nodata value (0.01) has been automatically
changed to (0.009999999776482582) to match the dtype of the data.
``` 
So the conversion is already happening so I think we might as well be explicit about it. I think it also has to do with the fact that `float` has built in for `inf` and `nan` whereas `int` doesn't which is causing these implicit conversions in a lot of unexpected places I think (see below) 

```Python
Python 3.11.3 | packaged by conda-forge | (main, Apr  6 2023, 08:57:19) [GCC 11.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np

In [2]: type(np.nan)
Out[2]: float

In [3]: np.float64('inf')
Out[3]: inf

In [4]: np.float64('nan')
Out[4]: nan

In [5]: np.int32('inf')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[5], line 1
----> 1 np.int32('inf')

ValueError: invalid literal for int() with base 10: 'inf'

In [6]: np.int32('nan')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[6], line 1
----> 1 np.int32('nan')

ValueError: invalid literal for int() with base 10: 'nan'

In [7]:
```

Reply:
You can convert for float but I prefer to have the error for int. Some models have strict requirements on data types so I prefer the user to have a clear error message and set a correct nodata value rather than converting the dtype of the data.

Reply:
After giving it some more thought, it seems that it's mostly the native python types that are not being very nice, so I've added some checks to convert those into the appropriate numpy types and otherwise just leave it as is, I think that should be a good compromise